yktrace/src/tir.rs

```diff
@@ -36,18 +36,96 @@ lazy_static! {
 }
 
 /// A TIR trace is conceptually a straight-line path through the SIR with guarded speculation.
+#[derive(Debug)]
 pub struct TirTrace {
-    _stmts: Vec<_TirOp>
+    ops: Vec<TirOp>
 }
 
 impl TirTrace {
-    pub(crate) fn new(_trace: &'_ dyn SirTrace) -> Self {
-        // FIXME: Use the SIR_MAP to convert the SirTrace to a TirTrace.
-        unimplemented!()
+    /// Create a TirTrace from a SirTrace.
+    /// Returns Err if a DefId is enountered for which no SIR is available. In the error case, the
+    /// trace built thus-far is returned inside the error.
+    ///
+    /// FIXME: This is not the final "intended" API. Noramlly we wouldn't care about the trace
+    /// built thus-far in an error case, but since we have no sensible way to stop the tracer at
+    /// the right time, inevitably the trace shoots off into code which has no SIR. In turn this
+    /// invalidates the trace. We only return the trace built so far for now so that we have
+    /// *something* to test in the interim.
+    pub(crate) fn new(trace: &'_ dyn SirTrace) -> Result<Self, Self> {
+        let mut ops = Vec::new();
+        let num_locs = trace.len();
+
+        for blk_idx in 0..num_locs {
+            let loc = trace.loc(blk_idx);
+            let body = match SIR_MAP.get(&DefId::from_sir_loc(loc)) {
+                Some(b) => b,
+                _ => return Err(Self { ops })
```

Comment:
Can this be anything other than `None`?

Reply:
It returns `Some` if there is an entry in the SIR map, but that can't be what you are asking?

Reply:
Let me rephrase: let's use `None` instead of `_` if that's the only thing the `_` can match.

Reply:
Why? The type system will ensure we are exhaustively matching.

Reply:
That statement makes little sense when you use the wildcard `_` operator: by definiton you've exhaustively matched everything.

Reply:
So think you are asking me to change `_` to `None`. We can, but it doesn't seem necessary. I won't argue over it :P

Reply:
Yes, we should use `None`. In general, `_` is something that one should be very wary of in production code (though in tests I'm much less worried about it).

Reply:
As you wish, but I'd have though it was obvious that `_` can only be `None` since the previous match was `Some`.

Reply:
What happens if the type you return changes in some way? The problem with `_` is that it allows you to sidestep most of the goodies of the type system, and that tends to bite you on the rear end years down the line during refactoring...

Reply:
OK, I take your point.